rasa/core/featurizers/precomputation.py

```diff
@@ -0,0 +1,462 @@
+        # given substate
+        key_attributes = set(sub_state.keys()).intersection(self.KEY_ATTRIBUTES)
+        for key_attribute in key_attributes:
+            key = self._build_key(sub_state={key_attribute: sub_state[key_attribute]})
+            message = self._table.get(key)
+            if not message:
+                raise ValueError(
+                    f"Unknown key {key}. Cannot retrieve features for substate "
+                    f"{sub_state}"
+                )
+            features_from_message = Features.groupby_attribute(
+                message.features, attributes=attributes
+            )
+            for feat_attribute, feat_value in features_from_message.items():
+                existing_values = features.get(feat_attribute)
+                # Note: the follwing if-s are needed because if we specify a list of
+                # attributes then `features_from_message` will contain one entry per
+                # attribute even if the corresponding feature list is empty.
+                if feat_value and existing_values:
+                    raise RuntimeError(
+                        f"Feature for attribute {feat_attribute} has already been "
+                        f"extracted from a different message stored under a key "
+                        f"in {key_attributes} "
+                        f"that is different from {key_attribute}. This means there's a "
+                        f"redundancy in the lookup table."
+                    )
+                if feat_value:
+                    features[feat_attribute] = feat_value
+        return features
+
+    def lookup_message(self, user_text: Text) -> Message:
+        """Returns the lookup table entry identified by the given user text.
+
+        Args:
+          user_text: the text of a user utterance
+        Raises:
+          `ValueError` if there is no message associated with the given user text
+        """
+        key = self._build_key({TEXT: user_text})
+        message = self._table.get(key)
+        if message is None:
+            raise ValueError(f"Expected a message with key {key} in lookup table.")
+        return message
+
+    def derive_messages_from_domain_and_add(self, domain: Domain) -> None:
+        """Adds all lookup table entries that can be derived from the domain.
+
+        Args:
+          domain: the domain from which we extract the substates
+          lookup_table: lookup table to which the substates will be added (as messages)
+        """
+        action_texts = domain.action_texts
+        action_names = domain.action_names_or_texts[
+            slice(0, -len(domain.action_texts) if domain.action_texts else None)
+        ]
+        if set(action_texts).intersection(action_names):
+            raise NotImplementedError(
+                "We assumed that domain's action_names_or_texts contains all "
+                "action names followed by the texts. Apparently that changed. "
```

Comment:
Guess no one has an interest in swapping those lists in the concatenation because I guess that would affect lots of label to index mappings downstream but the above looks so weird that I thought it's good to add this sanity check as explanation.

Reply:
Makes sense 👍🏻 Can still give some hints how to (maybe) resolve it?

Reply:
yep, will do